File: apps/android/app/src/main/java/com/exponential/app/domain/WebLinks.kt
package com.exponential.app.domain

/**
 * Builds the canonical web-app URL for sharing an issue. Mirrors the web route
 * shape and the iOS `WebLinks` helper:
 *   issue  {base}/t/{team}/boards/{board}/issues/{identifier}
 *
 * `base` is the account's instance URL; its trailing slash is trimmed so the
 * joined path never doubles up. All slugs/identifiers are synced locally, so
 * this composes without a network call. (Board-level sharing was removed —
 * sharing is issue-only across all clients.)
 */
object WebLinks {
    fun issueUrl(
        base: String,
        teamSlug: String,
        boardSlug: String,
        identifier: String,
    ): String = "${base.trimEnd('/')}/t/$teamSlug/boards/$boardSlug/issues/$identifier"

    /** A web-app URL the native app can render (EXP-92 App Links). */
    sealed interface Parsed {
        data class IssueRef(
            val teamSlug: String,
            val boardSlug: String,
            val identifier: String,
        ) : Parsed

        data class Invite(val token: String) : Parsed
    }

    /**
     * Inverse of [issueUrl] plus the invite route — the only two shapes the
     * manifest's autoVerify filter claims. Takes the already-decoded path
     * (Uri.getPath) rather than a Uri so it stays JVM-unit-testable; empty
     * segments are dropped (trailing slash tolerated) and deeper paths fail
     * the exact-length match — the app only claims what it can render.
     *
     * Only the current `/t/{team}/boards/{board}/issues/{id}` form is accepted
     * (the great rename, EXP-180): the web serves no `/w/` or `/projects/`
     * routes anymore — those legacy shapes are dead, so parsing them would
     * claim links the app can no longer resolve.
     */
    fun parsePath(path: String?): Parsed? {
        val parts = path.orEmpty().split('/').filter { it.isNotEmpty() }
        return when {
            parts.size == 6 && parts[0] == "t" &&
                parts[2] == "boards" && parts[4] == "issues" ->
                Parsed.IssueRef(parts[1], parts[3], parts[5])
            parts.size == 2 && parts[0] == "invite" -> Parsed.Invite(parts[1])
            else -> null
        }
    }
}

File: apps/ios/ExpCore/Sources/Domain/WebLinks.swift
import Foundation

/// Builds shareable web URLs into the running instance, mirroring the web app's
/// route shape:
///   issue:  {base}/t/{teamSlug}/boards/{boardSlug}/issues/{identifier}
///
/// EXP-180 (the great rename): the web's canonical routes are `/t/…/boards/…`
/// and the legacy `/w/` + `/projects/` forms are DEAD server-side (no
/// redirects), so this builder emits ONLY the new form and the parser resolves
/// ONLY the new form — a dead route must not deep-link into the app either.
///
/// `base` is the account's `instanceUrl` (per `AuthRepository`); any trailing
/// slash is trimmed first (precedent: TeamRepositoriesSection.webSettingsURL).
/// Slug/identifier path segments are percent-encoded defensively even though the
/// server only ever mints URL-safe slugs. (Board-level links were removed —
/// sharing is issue-only on every client.)
public enum WebLinks {
    /// Normalize an instance base URL string into a trailing-slash-free base.
    /// Returns nil for an empty/whitespace-only input.
    public static func normalizedBase(_ instanceUrl: String?) -> String? {
        guard let raw = instanceUrl?.trimmingCharacters(in: .whitespacesAndNewlines),
              !raw.isEmpty else { return nil }
        return raw.hasSuffix("/") ? String(raw.dropLast()) : raw
    }

    /// `{base}/t/{teamSlug}/boards/{boardSlug}/issues/{identifier}`
    public static func issue(
        instanceUrl: String?, teamSlug: String, boardSlug: String, identifier: String
    ) -> URL? {
        guard let base = normalizedBase(instanceUrl) else { return nil }
        let team = encode(teamSlug)
        let board = encode(boardSlug)
        let id = encode(identifier)
        return URL(string: "\(base)/t/\(team)/boards/\(board)/issues/\(id)")
    }

    private static func encode(_ segment: String) -> String {
        segment.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? segment
    }

    /// A web-app URL the native app can render (EXP-92 Universal Links) — the
    /// inverse of `issue(...)` plus the invite route. Only these two shapes are
    /// claimed in the associated-domains AASA; anything else returns nil.
    public enum Parsed: Equatable, Sendable {
        case issue(teamSlug: String, boardSlug: String, identifier: String)
        case invite(token: String)
    }

    /// Parse `{base}/t/{team}/boards/{board}/issues/{identifier}` plus
    /// `{base}/invite/{token}`. Splitting `url.path` (already percent-decoded)
    /// on "/" drops empty segments, so a trailing slash is tolerated while
    /// deeper paths (e.g. an issue's sub-tab) fail the exact-length match —
    /// deliberate: the app should only claim what it can render. The legacy
    /// `/w/` + `/projects/` forms are dead on the web (no redirects), so they
    /// deliberately return nil here too.
    public static func parse(_ url: URL) -> Parsed? {
        let parts = url.path.split(separator: "/").map(String.init)
        if parts.count == 6, parts[0] == "t",
           parts[2] == "boards", parts[4] == "issues" {
            return .issue(teamSlug: parts[1], boardSlug: parts[3], identifier: parts[5])
        }
        if parts.count == 2, parts[0] == "invite" {
            return .invite(token: parts[1])
        }
        return nil
    }
}
